File: Windows - console/Memory.cpp
#include "Memory.h"
#include "Converter.h"

const int SIZE_OF_MEMORY(65535);

Memory::Memory(){
	this->clear();
}

void Memory::clear(){
	for (int i = 0; i < SIZE_OF_MEMORY; i++)
		this->memory[i] = nullptr;
}

MemoryItem* Memory::get(int position){
	this->check(position);
	
	if (this->memory[position].get() == nullptr)
		throw std::out_of_range("Break the sequence of addresses");

	return this->memory[position].get();
}

void Memory::set(int position, MemoryItem* item){
	this->check(position);

	this->memory[position] = std::shared_ptr<MemoryItem>(item);
}

bool Memory::is_memory_cell_empty(int position){
	this->check(position);
	return this->memory[position].get() == nullptr;
}

void Memory::set_name(int position, std::string name){
	this->check(position);

	this->names[position] = name;
}

std::string Memory::get_name(int position)const{
	this->check(position);
	
	if (this->names.find(position) == this->names.end())
		return Converter::to_hex(position, 4);
	else
		return this->names.find(position)->second;
}

void Memory::check(int position)const{
	if (position < 0 || position > SIZE_OF_MEMORY)
		throw std::exception("Wrong address");
}

File: Windows - GUI CS/Memory.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace LM_GUI_UP
{
    class Memory
    {
        const int SIZE_OF_MEMORY = 65535;
        private MemoryItem[] memory;
        private Dictionary<int, string> names;

        public Memory(){
            if (memory != null)
                this.ClearMemory();
            else
            {
                memory = new MemoryItem[SIZE_OF_MEMORY];
                names = new Dictionary<int, string>();
            }
        }
        #region Methods for memory
        public MemoryItem GetCell(int position){
            this.Check(position);

            if (this.memory[position] == null)
                throw new Exception();
            //Appealing to a non-existent adress

            return this.memory[position];
        }
        public void SetCell(int position, MemoryItem item)
        {
            this.Check(position);

            this.memory[position] = item;
        }
        public void ClearMemory()
        {
            for (int i = 0; i < SIZE_OF_MEMORY; i++)
                this.memory[i] = null;
        }
        public bool IsMemoryCellEmpty(int position)
        {
            this.Check(position);

            return this.memory[position] == null;
        }
        #endregion

        #region Methods for names
        public void SetName(int position, string name)
        {
            this.Check(position);

            this.names.Add(position, name);
        }

        public string GetName(int position)
        {
            this.Check(position);

            if (this.names[position] == null)
                return Converter.DecToHex(position, 4); // possible mistakes
            else
               return this.names[position];
        }
        #endregion
        public void Check(int position)
        {
            if (position < 0 || position > SIZE_OF_MEMORY)
                throw new Exception("");
            //Memory bounds violation
        }
    }
}
